agent/proxycfg-glue/service_http_checks.go

```diff
@@ -0,0 +1,39 @@
+package proxycfgglue
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/local"
+	"github.com/hashicorp/consul/agent/proxycfg"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// CacheHTTPChecks satisifies the proxycfg.HTTPChecks interface by sourcing
+// data from the agent cache.
+func CacheHTTPChecks(c *cache.Cache) proxycfg.HTTPChecks {
+	return &cacheProxyDataSource[*cachetype.ServiceHTTPChecksRequest]{c, cachetype.ServiceHTTPChecksName}
+}
+
+// ServerHTTPChecks satisifies the proxycfg.HTTPChecks interface.
+// It sources data from the server agent cache if the service exists in the local state, else
+// it is a no-op since the checks can only be performed by local agents.
+func ServerHTTPChecks(deps ServerDataSourceDeps, cacheSource proxycfg.HTTPChecks, localState *local.State) proxycfg.HTTPChecks {
+	return serverHTTPChecks{deps, cacheSource, localState}
+}
+
+type serverHTTPChecks struct {
+	deps        ServerDataSourceDeps
+	cacheSource proxycfg.HTTPChecks
+	localState  *local.State
+}
+
+func (c serverHTTPChecks) Notify(ctx context.Context, req *cachetype.ServiceHTTPChecksRequest, correlationID string, ch chan<- proxycfg.UpdateEvent) error {
+	// if the service exists in the server agent local state, delegate to the cache data source
+	if c.localState.ServiceExists(structs.ServiceID{ID: req.ServiceID, EnterpriseMeta: req.EnterpriseMeta}) {
+		return c.cacheSource.Notify(ctx, req, correlationID, ch)
+	}
```

Comment:
Service IDs are technically only unique within a node. So it's possible (albeit unlikely) that the xDS stream is for a service with the same ID registered to a different node, and returning this node's checks would be incorrect.

I think for completeness we should probably extend the `ServiceHTTPChecksRequest` to include the node name and compare it to the local node name.

Sorry! I should've mentioned this before.

Reply:
Makes sense!  Updated!